### DISCUSSION TITLE
Using typescript with xstate

### DISCUSSION THREAD

**Discussion by juan267** [statelyai/xstate#1719] [Q&A] [answered]
Hi, thanks for the great library!

One of the few hiccups that I found when I started writing my first state machines where that using Typescript with it was not so straightforward. The docs are good but lack a complete example that uses actions with arguments, so I end up spending a lot of time looking at examples scattered around the web to correctly typed my machine. In the end I think I finally made it work. this is the Auth machine that I am currently using: 

```js
import { createMachine, assign, interpret, MachineOptions, MachineConfig, StateNode } from 'xstate'
import { isTokenValid } from '/@/auth/tokenManager'
import { fetchUser } from '/@/api/axios/models/user'
import { logInUser, logOutUser } from '/@/api/axios/auth'
import { User } from '/@/graphql-operations'
import { router } from '/@/router'

interface AuthContext {
  user: Nullable<User>
  login: {
    email: string
    password: string
  }
}

type LOG_OUT_EVENT = { type: 'LOG_OUT' }
type LOG_IN_FORM_SUBMIT_EVENT = {
  type: 'LOG_IN_FORM.submit'
  login: { email: string; password: string }
}
type FETCH_USER_EVENT = { type: 'FETCH_USER'; data: { data: { data: User } } }

type AuthEvent = LOG_OUT_EVENT | LOG_IN_FORM_SUBMIT_EVENT | FETCH_USER_EVENT

type AuthTypedStates =
  | {
      value: 'initAuth'
      context: AuthContext & {
        user: null
      }
    }
  | {
      value: 'fetchingUser'
      context: AuthContext & {
        user: null
      }
    }
  | {
      value: 'loggingInUser'
      context: AuthContext & {
        user: null
      }
    }
  | {
      value: 'loggedIn'
      context: AuthContext & {
        user: User
      }
    }
  | {
      value: 'loggedOut'
      context: AuthContext & {
        user: null
      }
    }

interface AuthStates {
  states: {
    initAuth: StateNode
    fetchingUser: StateNode
    loggingInUser: StateNode
    loggedIn: StateNode
    loggedOut: StateNode
  }
}

const authMachineConfig: MachineConfig<AuthContext, AuthStates, AuthEvent> = {
  context: {
    user: null,
    login: {
      email: '',
      password: '',
    },
  },
  initial: 'initAuth',
  states: {
    initAuth: {
      always: [
        {
          cond: 'shouldFetchUser',
          target: 'fetchingUser',
        },
        {
          target: 'loggedOut',
        },
      ],
    },
    fetchingUser: {
      invoke: {
        id: 'fetchUser',
        src: 'fetchUser',
        onDone: {
          target: 'loggedIn',
          actions: 'assignUser',
        },
        onError: {
          target: 'loggedOut',
        },
      },
    },
    loggingInUser: {
      invoke: {
        id: 'logInUser',
        src: 'logInUser',
        onDone: {
          target: 'loggedIn',
          actions: 'assignUser',
        },
        onError: { target: 'loggedOut' },
      },
    },
    loggedIn: {
      on: {
        LOG_OUT: {
          actions: ['logOutUser'],
          target: 'loggedOut',
        },
      },
    },
    loggedOut: {
      exit: ['goToAccountIndex'],
      on: {
        'LOG_IN_FORM.submit': {
          actions: 'assignLogin',
          target: 'loggingInUser',
        },
      },
    },
  },
}

const authMachineOptions: Partial<MachineOptions<AuthContext, AuthEvent>> = {
  actions: {
    assignUser: assign({
      user: (_, event) => {
        if (event.type === 'FETCH_USER') {
          return event.data.data.data
        } else {
          return null
        }
      },
    }),
    assignLogin: assign({
      login: (_, event) => {
        if (event.type === 'LOG_IN_FORM.submit') {
          return event.login
        } else {
          return { email: '', password: '' }
        }
      },
    }),
    goToAccountIndex: () => {
      router.push({ path: '/account' })
    },
    logOutUser,
  },
  guards: {
    shouldFetchUser: () => !document.location.pathname.includes('login') && isTokenValid(),
  },
  services: {
    fetchUser,
    logInUser: (ctx) => logInUser(ctx.login),
  },
}

const authMachine = createMachine<AuthContext, AuthEvent, AuthTypedStates>(
  authMachineConfig,
  authMachineOptions
)

export const authMachineService = interpret(authMachine, { devTools: true }).start()
```

Some things that where confusing to me: 

- only `createMachines` accepts the generic for TypedStates, MachineConfig needs to have the regular States type. Reading the docs I good the impression that `TypedStates` where the new thing and States was not necessary, but in reality you need both? 
- You need to use type guards eg ` if (event.type === 'LOG_IN_FORM.submit') {` to correctly have the action event typed. For some reason I though that the action would know which event would be using it. And doing something like this ` login: (_, event: LOG_IN_FORM_SUBMIT_EVENT) => {` as the action type makes typescript unhappy
- you need to do Partial<MachineOptions> to not have to pass all the keys inside MachineOptions

I just wanted to point out this aspects to ask the community if my implementation looks alright or if it can be improved. Also if you think this `AuthMachine` is a good example of a more complete typed machine I think it would be nice to add something like it to the docs. I would be happy to make such PR 😄

**Comment by davidkpiano** [comment, marked as answer]
We're working hard to improve the TypeScript experience in XState; it definitely hasn't been easy! A lot of things we want to do with strongly typing machines is difficult (if not impossible) to do with TypeScript alone. This is why the [XState-CodeGen](https://github.com/mattpocock/xstate-codegen) project by @mattpocock is really useful and important. Give it a try!

> only createMachines accepts the generic for TypedStates, MachineConfig needs to have the regular States type. Reading the docs I good the impression that TypedStates where the new thing and States was not necessary, but in reality you need both?

The `StateSchema` generic type is not necessary. Honestly, I usually pass `any` for this. Ideally (and in the future), the `StateSchema` can be directly inferred from the machine config itself.

> You need to use type guards eg if (event.type === 'LOG_IN_FORM.submit') { to correctly have the action event typed. For some reason I though that the action would know which event would be using it. And doing something like this login: (_, event: LOG_IN_FORM_SUBMIT_EVENT) => { as the action type makes typescript unhappy

In reality, there are two "types" of actions, as far as TypeScript is concerned:

- **General actions**, which need to assume they can be called as a result of _any_ event. These actions occur as:
  - `entry` actions
  - `exit` actions
  - actions in `always` transitions
  - actions in wildcard transitions (`on: { '*': { ... } }`)
- **Restricted actions**, which must be restricted for the specific event in which they are called:
  - transition `actions` are the only actions where this applies.

We're currently working on improving the typing here (allowing "general actions" to be used where "restricted actions" are expected), but for now, you will need to assert the event is the expected type, unless the action is defined inline, in which I believe TypeScript will automatically infer those actions correctly.

> you need to do Partial to not have to pass all the keys inside MachineOptions

Same thing as above applies: if defining the machine options inline, it'll be typed for you. If defining them externally, it won't, and you need to be explicit. This is not unique to XState; inference is your friend, but only if what you want to infer is in the place where it can be readily inferred.

Some of the progress is on this WIP PR: https://github.com/davidkpiano/xstate/pull/1670 - we want to make as much inferrable as possible, and also make things like actions, guards, etc. strongly typed:

```ts
createMachine({
  schema: {
    events: createSchema<SomeEvents>(),
    actions: createSchema<SomeActions>(),
    guards: createSchema<SomeGuards>()
  }
});
```

And we also want to make it so that you can pass anything (like JSON Schema) into `createSchema<SomeEvents>(eventsSchema)` so that you can use those same types/schemas in runtime, as well.

Still a work-in-progress! 🚧

**Comment by juan267** [comment]
@davidkpiano Thanks a lot for the pointers! both XState-CodeGen  and #1670 look very promising 😄

**Comment by EightArmCode** [comment]
> I end up spending a lot of time looking at examples scattered around the web to correctly typed my machine.

Same here! Thank you for the example and sharing your findings.
